File: Python/0349-e-intersection-of-two-arrays.py
"""
Set
Time: O(m + n)
Space: O(m + n)
"""
class Solution_1:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        set1 = set(nums1)
        set2 = set(nums2)

        if len(set1) < len(set2):
            return self.set_intersection(set1, set2)
        else:
            return self.set_intersection(set2, set1)

    def set_intersection(self, set1, set2):
        return [x for x in set1 if x in set2]

"""
Hash Table
Time: O(m + n)
Space: O(m + n)
"""
class Solution_2:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        m, n = len(nums1), len(nums2)
        hash = {}
        if m > n:
            nums1, nums2 = nums2, nums1
        for num in nums1:
            hash[num] = True
        res = []
        for num in nums2:
            if num in hash and hash[num]:
                res.append(num)
                hash[num] = False
        return res

"""
Merge two sorted arrays
Time: O(nlogn + mlogm)
Space: O(1)
"""
class Solution_3:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        res = set()
        p1, p2 = 0, 0
        while p1 < len(nums1) and p2 < len(nums2):
            if nums1[p1] == nums2[p2]:
                res.add(nums1[p1])
                p1 += 1
                p2 += 1
            elif nums1[p1] < nums2[p2]:
                p1 += 1
            else:
                p2 += 1
        return list(res)

"""
Binary Search
Time: O(nlogn + mlogm)
Space: O(1)
"""

File: leetcode/Java/0349-e-intersection-of-two-arrays.java
/*
 * Set
 */
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set1 = new HashSet<Integer>();
        for (Integer n : nums1)
            set1.add(n);
        Set<Integer> set2 = new HashSet<Integer>();
        for (Integer n : nums2)
            set2.add(n);

        if (set1.size() < set2.size())
            return set_intersection(set1, set2);
        else
            return set_intersection(set2, set1);
    }

    public int[] set_intersection(Set<Integer> set1, Set<Integer> set2) {
        int[] output = new int[set1.size()];
        int i = 0;
        for (Integer s : set1)
            if (set2.contains(s))
                output[i++] = s;
        return Arrays.copyOf(output, i);
    }
}